File: lib/hacks/each_instance_variable.rb
# Ruby Treasures 0.4
# Copyright (C) 2002 Paul Brannan <paul@atdesk.com>
# 
# You may distribute this software under the same terms as Ruby (see the file
# COPYING that was distributed with this library).
# 
require 'each_instance_variable.so'

# The following methods are implemented in C:
if false then

  class Object
  private
    ##
    # Iterate over every instance variable in the given object, yielding
    # [ name, obj ], where name is the Symbol representing the name of
    # the instance variable, and obj is the instance variable itself.
    #
    # This is useful for operations such as deep freeze and deep copy.
    # It can also come in handy for writing custom marshallers.  A less
    # common use would be to access instance variables whose name do not
    # begin with the @ character (such as the Exception class's mesg
    # variable).
    #
    def each_instance_variable
    end
  end
end

File: ext/hacks/each_instance_variable.c
/*
 * Ruby Treasures 0.4
 * Copyright (C) 2002 Paul Brannan <paul@atdesk.com>
 * 
 * You may distribute this software under the same terms as Ruby (see the file
 * COPYING that was distributed with this library).
 * 
 */
#include <ruby.h>
#include <intern.h>

VALUE ruby_each_instance_variable(VALUE self) {
    int j;
    VALUE instance_vars = rb_obj_instance_variables(self);
    VALUE yield_value = rb_ary_new2(2);
    struct RArray * instance_vars_ary = RARRAY(instance_vars);
    struct RArray * yield_value_ary = RARRAY(yield_value);
    ID id;
    yield_value_ary->len = 2;
    
    for(j = 0; j < instance_vars_ary->len; ++j) {
        id = rb_intern(STR2CSTR(instance_vars_ary->ptr[j]));
        yield_value_ary->ptr[0] = ID2SYM(id);
        yield_value_ary->ptr[1] = rb_ivar_get(self, id);
        rb_yield(yield_value);
    }

    return Qnil;
}

void Init_each_instance_variable(void) {
    rb_define_private_method(rb_cObject, "each_instance_variable", ruby_each_instance_variable, 0);
}

